pyat/at/lattice/lattice_object.py

```diff
@@ -436,6 +454,70 @@ def particle(self, particle: Union[str, Particle]):
         else:
             self._particle = Particle(particle)
 
+    def set_fillpattern(self, bunches=1):
+        """Function to generate the filling pattern lof the ring.
+        The filling pattern is computed as:
+
+        ``bunches/numpy.sum(bunches)``
+
+        This function also generates the bunch spatial distribution
+        accessible with ``Lattice.bunch_spos``
+
+        Keyword Arguments:
+           bunches:  scalar or array of positive double or bool to define
+                     the bunch distribution.
+                     For scalar input, equidistant bunches are assumed.
+                     ``ring.harmonic_number`` has to be a multiple of
+                     ``bunches``.
+                     For array input the condition
+                     ``len(bunches)==ring.harmonic_number`` is required.
+                     (default=1, single bunch configuration).
+        """
+
+        if numpy.isscalar(bunches):
+            if self.harmonic_number % bunches == 0:
+                fp = numpy.ones(bunches)
+                bs = self.circumference/bunches
+                fs = bs*numpy.arange(bunches)
+            else:
+                raise AtError('Harmonic number has to be a '
+                              'multiple of the scalar input '
+                              'bunches')
+        else:
+            assert len(bunches) == self.harmonic_number, \
+                'bunches array input has to be of shape ({0},)' \
+                .format(self.harmonic_number)
+            assert numpy.all(bunches >= 0.0), \
+                'bunches array can contain only positive numbers'
+            fp = numpy.ones(self.harmonic_number)[bunches>0.0]
```

Comment:
If you want to keep the bunch charge, you need:
```
fp = bunches[bunches > 0]

Reply:
Done